tests/unit/src/test/scala/tests/FoldingRangeSuite.scala

```diff
@@ -0,0 +1,68 @@
+package tests
+
+import java.nio.file.Paths
+import java.util.UUID
+import org.eclipse.lsp4j.FoldingRange
+import tests.MutableText.Insert
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.FoldingRangeProvider
+import scala.meta.io.AbsolutePath
+
+object FoldingRangeSuite extends CustomInputExpectSuite("foldingRange") {
+  private val buffers = Buffers()
+  private val trees = TestingTrees(buffers)
+
+  private val foldingRangeProvider =
+    new FoldingRangeProvider(trees, foldOnlyLines = false)
+
+  override protected def obtainFrom(file: InputFile): String = {
+    val scalaSource = file.input.text
+      .replaceAll(">>\\w+>>", "")
+      .replaceAll("<<\\w+<<", "")
+
+    val actualRanges = findFoldingRangesFor(scalaSource)
+    generateOutput(scalaSource, actualRanges)
+  }
+
+  private def findFoldingRangesFor(source: String): Seq[FoldingRange] = {
+    import collection.JavaConverters._
+    val path = registerSource(source)
+    foldingRangeProvider.getRangedFor(path).asScala
+  }
+
+  private def registerSource(source: String): AbsolutePath = {
+    val name = UUID.randomUUID().toString + ".scala"
+    val path = AbsolutePath(Paths.get(name))
+    buffers.put(path, source)
+    path
+  }
+
+  private def generateOutput(
+      scalaSource: String,
+      actualRanges: Seq[FoldingRange]
+  ): String = {
+    val mutableText = MutableText(scalaSource)
```

Comment:
Can we use `TextEdits.applyEdits(edits: List[TextEdit])`?

```scala
val insertions = for {
      range <- actualRanges
      insertion <- Seq(opening(range), closing(range))
    } yield insertion
TextEdits.applyEdits(insertions)

private def opening(range: FoldingRange): l.TextEdit =
    new l.TextEdit(
      new l.Range(...),
      s">>${range.getKind}>>"
    )
```

We might be able to avoid `MutableText` in that case.

Reply:
Note, by convention I try to use `import org.eclipse.{lsp4j => l}` and `l.Range` instead of `Range` (to avoid conflict with stdlib).

Reply:
`TextEdits` is part of the `cross` module - hence not visible in the `unit` module. I do not see any "shared test" module - what would be the best approach here? (because it does in fact seem to render the `MutableText` void)

Reply:
Is it possible to move `TextEdits` to `mtest/main`?